enhancements/etcd/automated-backups-no-config.md

```diff
@@ -0,0 +1,308 @@
+without provided configuration from the user.
+
+This document outlines the possible approaches that were investigated, discussed and tested, in addition to the final implemented approach.  
+
+## Motivation
+
+The current method of **Backup** and **Restore** of an Openshift cluster is outlined in  [Automated Backups of etcd](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md). 
+This approach relies on user supplied configuration, in the form of `config.openshift.io/v1alpha1 Backup` CR.
+
+To improve the user's experience, this work proposes taking etcd backup using default configurations from day one, without user interference.
+
+Moreover, for recovery in disaster scenarios, where more than one control-plane node is lost from an Openshift cluster, this proposal attempts to store backups on all control-plane nodes.
+Having backups of an Openshift cluster on all control-plane nodes, improves the recovery probability in disaster scenarios.
+
+### User Stories
+
+* As a cluster administrator, I would like an OCP cluster backups to be taken without user's supplied configuration from day one, after installation.
+* As a cluster administrator, I would like to an OCP cluster backups available on all master nodes, so that I have a recent cluster state to recover from in the event of losing control-plane nodes.
+
+### Goals
+
+* Backups should be taken without configuration after cluster installation from day 1.
+* Backups are saved to a default location on each master node's file system.
+* Backups are taken according to a default schedule.
+* Backups are being maintained according to a default retention policy.
+
+### Non-Goals
+
+* Save cluster backups to remote cloud storage (e.g. S3 Bucket).
+  - This could be a future enhancement or extension to the API.
+  - Having backups in an independent storage from OCP cluster is crucial for Disaster Recovery scenarios.
+* Automate cluster restoration.
+* Provide automated backups for non-self hosted architectures like Hypershift.
+
+## Proposal
+
+This work utilizes the [API](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups.md#api-extensions) from current `CronJob` based implementation.
+Moreover, it uses the same [PeriodicBackupController](https://github.com/openshift/cluster-etcd-operator/blob/master/pkg/operator/periodicbackupcontroller/periodicbackupcontroller.go#L81) as this proposal builds upon it, and also under the same feature gate.
+The default configuration is created through `config.openshift.io/v1alpha1 Backup` CR, with name `default`, to distinguish it from other user supplied `Backup` CRs.
+Using a CR enable users to override the default backup configuration.
+
+By design, the _**No-Config**_ backup approach works _orthogonal_ to the [Cron](https://github.com/openshift/enhancements/blob/master/enhancements/etcd/automated-backups) based method, and not a replacement.
```

Comment:
Before anything else we need to expand on this section to mention how exactly are the `automated local backup` and `automated PVC backup` orthogonal or related to each other. 
How does one get enabled and why they need to share the same CRD?

Elsewhere you've mentioned:
>However, creating a `Backup` CR with `name=default` enables the container.

So that means the local backups get enabled when you have a special name i.e `default`? The name isn't very indicative of how this chooses between the `local` vs `PV` storage location.

And the `Backup` CR also has a field for `PVCName`. So is it that just getting ignored if the user makes edits to that when the name is `default`? Or is it immutable in that case? 
And then if you create another `Backup` CR with `name=foobar` does that enable the automated backups to PVC feature?

So now you have 2 CRs of the same type/spec with different names for different behaviors? Seems a bit unintuitive and obscure to me. 

Which leads me to the question on whether or not both these features need to share the same CR.
If they do, then what are the possible locations where the backups are being saved 
1. Locally across all nodes
2. On the PVC 
3. On both at the same time, or exclusively on 1 or 2.

Instead of relying on a magic name to enable the `local` option can we not explicitly have an API field(or multiple) to make it clear what location is being picked and how that can be changed.

And if there are drawbacks to having a unified API for both options  then we should list them out, and explore what having an entirely separate CR for the `automated local backup` would look like to truly make it orthogonal.

Reply:
> Before anything else we need to expand on this section to mention how exactly are the automated local backup and automated PVC backup orthogonal or related to each other.
How does one get enabled and why do they need to share the same CRD if they're orthogonal?

Added 👍🏽

Reply:
> So that means the local backups get enabled when you have a special name i.e default? The name isn't very indicative of how this chooses between the local vs PV storage location.

So the `default` name has been selected by @tjungblu since the very beginning, and we agreed to use it as indicator for the `NoConfig` method 👍🏽

Reply:
> And the Backup CR also has a field for PVCName. So is that just getting ignored if the user makes edits to that when the name is default? Or is it immutable in that case?

Yes it is being ignored, it is never mounted see https://docs.google.com/document/d/1813mZo4KEfTWw_FRbmG8Un91oPA9-eF29FcrZk-RLxE/edit?usp=sharing

Reply:
> And then if you create another Backup CR with name=foobar does that enable the automated backups to PVC feature?
So now you have 2 CRs of the same type/spec with different names for different behaviors? Seems a bit unintuitive and obscure to me.


Creating any `Backup` CR with any other name than `default`, triggers the `Cron` based approach

Only CR with name `default` is being handled by the NoConfig 

see test here https://docs.google.com/document/d/1Ee9a7GJsae8EA7komPcEV72YlRgbSbR0B0KSo3XRBUc/edit?tab=t.0#heading=h.hyvb3n32a77z 

Also applying two consecutive Backup CRs with name = `default`, the latest CR is being used to configure the DaemonSet and the Pods

Reply:
> Instead of relying on a magic name to enable the local vs pvc option can we not explicitly have an API field(or multiple) to make it clear what location is being picked and how that can be changed.

Well, I think the purpose of this work, is to avoid having user's configuration. In this sense, adding an `API` while the user should not use it ?!!! 🤔

Reply:
I think we may want to get the three of us together to have a conversation about this project. 

One major question that still exists for me is, why is this not just a special case of the periodic backup.

It strikes me, that the CEO could enforce creating a `configv1.Backup` (from the periodic EP) that configures backing up to a hostpath on the master instances, and, prevent users from deleting or tweaking that in ways we don't want them to with admission time checks.

The default backup object that CEO creates would configure these default backups, but leverage the same cronjob based implementation as the rest of the backups.

Meaning that all that is needed for this project, is to create and work out the caveats of creating that default backup object, and not have to completely reimplement a way to handle the backups?

What about the existing periodic backup means it's not compatible in this way?

Reply:
> It strikes me, that the CEO could enforce creating a configv1.Backup (from the periodic EP) that configures backing up to a hostpath on the master instances, and, prevent users from deleting or tweaking that in ways we don't want them to with admission time checks.

Actually user could change the default configs by editing the `default` Backup CR.

> The default backup object that CEO creates would configure these default backups, but leverage the same cronjob based implementation as the rest of the backups.

I actually think the main reason we have changed the approach is to have replicas across all master nodes iiuc

> What about the existing periodic backup means it's not compatible in this way?

I think the purpose since the beginning was to have both approaches working simultaneously iiuc


> I think we may want to get the three of us together to have a conversation about this project.

I am fine, I will work out an invite for this 👍🏽

Reply:
> I actually think the main reason we have changed the approach is to have replicas across all master nodes iiuc

Currently, what does a periodic backup do? Just run the one replica on any node? Or does it run one replica on any master node?

> I think the purpose since the beginning was to have both approaches working simultaneously iiuc

I don't see why having multiple periodic backups configured would mean that there's a non simultaneous element

Reply:
> Currently, what does a periodic backup do? Just run the one replica on any node? Or does it run one replica on any master node?

I think it goes to any `master node` 

cc @tjungblu @hasbro17

Reply:
> I don't see why having multiple periodic backups configured would mean that there's a non simultaneous element

What I meant, is that both approaches exist :)